ledger/participant-integration-api/src/main/scala/platform/store/cache/StateCache.scala

```diff
@@ -98,17 +105,59 @@ private[platform] case class StateCache[K, V](
     registerUpdateTimer,
     pendingUpdates.synchronized {
       val validAt = cacheIndex
-      val eventualValue = Future.delegate(fetchAsync(validAt))
-      val pendingUpdatesForKey = pendingUpdates.getOrElseUpdate(key, PendingUpdatesState.empty)
-      if (pendingUpdatesForKey.latestValidAt < validAt) {
-        pendingUpdatesForKey.latestValidAt = validAt
-        pendingUpdatesForKey.pendingCount += 1
-        registerEventualCacheUpdate(key, eventualValue, validAt)
-          .flatMap(_ => eventualValue)
-      } else eventualValue
+      putPendingForKey(key, validAt, Future.delegate(fetchAsync(validAt)))
     },
   )
 
+  private def putPendingForKey(key: K, validAt: Offset, eventualValue: Future[V])(implicit
+      loggingContext: LoggingContext
+  ): Future[V] = {
+    val pendingUpdatesForKey = pendingUpdates.getOrElseUpdate(key, PendingUpdatesState.empty)
+    if (pendingUpdatesForKey.latestValidAt < validAt) {
+      pendingUpdatesForKey.latestValidAt = validAt
+      pendingUpdatesForKey.pendingCount += 1
+      registerEventualCacheUpdate(key, eventualValue, validAt)
+        .flatMap(_ => eventualValue)
+    } else eventualValue
+  }
+
+  /** Batch fetches many contracts at once to avoid incremental fetches to database */
+  def putAsyncMany(keys: Seq[K], fetchAsync: Offset => Future[Map[K, V]], empty: K => Future[V])(
+      implicit loggingContext: LoggingContext
+  ): Future[Unit] =
+    Future
+      .sequence(
+        Timed.value(
+          registerUpdateTimer,
+          pendingUpdates.synchronized {
+            val validAt = cacheIndex
+            val eventualValue = Future.delegate(fetchAsync(validAt))
+            keys.map { key =>
+              putPendingForKey(
+                key,
+                validAt,
+                eventualValue.map(_.get(key)).transformWith {
+                  case Success(Some(value)) => Future.successful(value)
+                  // the state cache expects the loader to be Future[V] instead of Future[Option[V]]
+                  // missing contracts are signalled via an exception which is semi ideal
+                  case Success(None) => empty(key)
+                  case Failure(exception) => Future.failed(exception)
+                },
+              ).map(_ => ())
+                .transformWith {
+                  // as the preloading shares 1:1 the code with the single contract loading,
+                  // we need to catch contract not found errors here and just ignore them as
+                  // otherwise we'll abort the interpretation prematurely
```

Comment:
so what's the problem if the interpretation fails prematurely? Are you trying to optimize for the case where the client does not wait for completion of previous commands (and the resolution can succeed a bit later)?